client/src/lib.rs

```diff
@@ -56,24 +56,49 @@ where
     CmdT: TpmCommand,
     T: Tpm,
 {
-    Ok(run_command_with_handles(cmd, CmdT::Handles::default(), CmdSessions::default(), tpm)?.0)
+    Ok(run_command_with_handles(cmd, CmdT::Handles::default(), (), tpm)?.0)
+}
+
+fn marshal_auth_size(auth_offset: usize, buffer: &mut [u8]) -> TssResult<usize> {
+    let auth_size = (auth_offset - size_of::<u32>()) as u32;
+    auth_size.try_marshal(buffer)?;
+    Ok(auth_offset)
 }
 
 /// Adds any command sessions to the command buffer.
-pub fn write_command_sessions(sessions: &CmdSessions, buffer: &mut [u8]) -> TssResult<usize> {
+pub fn write_command_sessions<
+    X: Session,
+    Y: Session,
+    Z: Session,
+    AA: AuthorizationArea<X, Y, Z>,
+>(
+    sessions: &AA,
+    buffer: &mut [u8],
+) -> TssResult<usize> {
     if sessions.is_empty() {
         return Ok(0);
     }
     let mut auth_offset = size_of::<u32>();
-    for session in sessions {
-        // TODO: Support parameter encryption.
-        auth_offset += session
-            .get_auth_command()
-            .try_marshal(&mut buffer[auth_offset..])?;
-    }
-    let auth_size = (auth_offset - size_of::<u32>()) as u32;
-    auth_size.try_marshal(buffer)?;
-    Ok(auth_offset)
+    let (s1, s2, s3) = sessions.decompose_ref();
+    let Some(s1) = s1 else {
```

Comment:
Technically these could be deduped by using dynamic dispatch;

```rs
impl AuthorizationArea {
  fn iter(&self) -> impl Iterator<Item=&dyn Session>;
}
```

```rs
for session in sessions.iter() {
  auth_offset += s1
    .get_auth_command()
    .try_marshal(&mut buffer[auth_offset..])?;
}

// Probably don't need a separate function for this
// if we dedupe the code using dyn references.
marshal_auth_size(auth_offset, buffer)
```

I know you're trying to avoid that, but I'm not totally sure it's worth avoiding it myself. So I figured I'd at least say there's an alternative way to organize this.

Reply:
So I have not tried it but I think it should be

```rust
fn as_array(&self) -> &[&dyn Sessions];
```
or 
```rust
fn as_array(&self) -> Vec[&dyn Sessions];
```
because they do not have to be 3 sessions, they can be up to 3. I prefer to avoid allocation whenever possible, but the first one would not work. The problem being this array cannot be created inside `as_array`.

That leaves us with

```rust
fn into_slice(&self, buf: &mut [&dyn Sessions; 3]) -> &[&dyn Sessions];
```
But that needs lots of infrastructure almost as much as the duplicated code, and there is the problem of how does `buf` gets initialized. But I hear you, and I understand that in some cases dynamic dispatch can make things more convenient to work with. But, it adds overhead and sometimes complications:
1. The overhead arises from the fact that there is dynamic dispatching logic (last I checked on microbenchmark it was ~15x slower in terms of runtime), and it is almost always the case that you would need some form of heap allocation. 
2. The complications arise from having to worry about object safety and that that we will never be able to do thing like `dyn A + B` and we will have to things along the lines of `trait C : A + B .....` and then we worry again about object safety stuff.


But I would say the following, it is much much easier to move from static traits to dynamic dispatcher, because we are simply throwing away type information. But it will be much harder to do it the other way around, and in some cases we might need to restructure the code, because the type information has to come from somewhere. So How wold you feel about deferring the decision to moving to dynamic dispatcher until we really have no choice? I am just concerned that the moment we switch to it, it we will definitely have no choice of going back easily.

Reply:
> it is almost always the case that you would need some form of heap allocation.

That's not true at all: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9de66e14d751bd6b72afa868b439f43a

You don't need heap allocation for dynamic dispatch, and I strongly doubt the claim of 15x slower. I would want to see some data around that, and also I'd want to see about heap-allocated structures vs non-heap allocated (my proposal doesn't require heap allocation).

> The complications arise from having to worry about object safety

I don't understand why everyone is so scared of object safety... It's not that hard, it's basically a name for "things the compiler literally cannot do". Let's take it to a C++ context for a comparison:

```cpp
class Interface {
 public:
  // Doesn't make sense; can't dynamic dispatch on a non-object
  static virtual int function() = 0;
};
```

This is the same as:

```rs
trait Interface {
  // You can form this trait, but it's not "object safe".
  // So you can't call this from a `&dyn Interface`.
  fn function() -> i32;
}
```

Another example, you can't dynamic dispatch on a template function:

```cpp
class Interface {
  // Doesn't make sense, how do I build a vtable
  // with an unknown number of functions?
  virtual int function<T>() = 0;
}
```

This is the same as:

```rs
trait Interface {
  // You can create this, but it's not "object safe",
  // for the same reasons as the C++ example.
  fn function<T>() -> i32;
}
```

> will never be able to do things like `dyn A + B`

So? Where will we need that? I say let's cross that bridge when we come to it.

The reason that isn't allowed is because Rust doesn't want to magically construct the vtable of `A + B` for you, when you use supertraits you expressly say "I want this vtable". I actually like that better than implicit vtables being constructed all over the place.

Reply:
Also, sorry I meant something more like this:

```rs
impl AuthorizationArea {
  fn as_array(&self) -> [Option<&dyn Sessions>; 3];
}
```

Then you can `array.iter().filter_map(std::convert::identity)` and only handle things that are set to `Some`. I definitely don't want to allocate for this purpose.

Anyways, it's only 3 things, so maybe not worth it, but I do think it's possible (without allocation or tricky memory bits).

Reply:
Oh, I didn't answer one other thing:

> So How wold you feel about deferring the decision to moving to dynamic dispatcher until we really have no choice?

Well, as I said in the issue, I'm not the biggest fan of this path - but I'm okay with us doing it and seeing where it takes us. I could be totally wrong about it, maybe it'll be just fine, and I'd rather us make some progress than me continue to block on "perfect".

So I'm not blocking on dynamic dispatch - but I strongly recommend it.

Reply:
> and I strongly doubt the claim of 15x slower. I would want to see some data around that

I did micro-benchmarks quite some time ago,  but other did the same too. Quick googling lead me to this post https://medium.com/digitalfrontiers/rust-dynamic-dispatching-deep-dive-236a5896e49b which shows dynamic dispatching has overhead of 3.3X (330%) in micro benchmarks. it is not 15X but I think it depends on the benchmark and definitely  rustc optimizations is improving but it is still can be notable difference especially for a microcontroller class chip.

i will fix other issues, but as for this one, I slightly prefer avoiding dyn until last minute, it will always be extremely easy to switch once we have real justification.

Reply:
It depends on what you're measuring. If it's how to get to the function that you want to execute, sure ~3x sounds right because you have a double-indirection through the vtable.

But after you get to the function, it's not like, magically slower because it's a dynamic function or anything like that. The worst part is you can't inline the code so some optimizations won't happen, but I still even question 3x slowdown, because really it depends on what kind of function you're measuring and what it does and all that.

I disagree, but I won't block.

Reply:
> I disagree, but I won't block.

To be clear on this, I'll resolve this comment thread and approve. I think you were going to make some other changes, but nothing else worth withholding approval on.